### ISSUE TITLE
External Include Paths Point into Unstable Execroot Instead of Stable External Directory

### ISSUE THREAD

**Issue by cpsauer** [bazelbuild/intellij#2766]
Hello wonderful Bazel folks,

**_TL;DR / summary:_** C++ header finding--and probably more--breaks for external libraries when you run a build command that uses a different set of external libraries. This issue can be fixed by having paths go through `<outputBase>/external/`, rather than the current `<outputBase>/execroot/__main__/external/`.

**_Discovery context:_** I'd noticed this after initially making the same mistake when writing a VSCode<>Clangd<>Bazel integration. I think this has the same root cause as https://github.com/bazelbuild/intellij/issues/1256 --a deeper fix here might allow removal of the code that patched it in https://github.com/bazelbuild/intellij/commit/81afc2cd1212828f33531d2705ce61172cfd237e. A similar problem has also caused issues in Tulsi.

**_What's going wrong in more detail:_**
When the plugin gets paths to external libraries from Bazel, they usually point into the execution root's external folder, the temporary sandbox of symlinks reconfigured for each build. This leads leads the plugin to point the IDE into the temporary sandbox for external header search paths (and other things, I'd assume). 

That's a trap. When you run a build that doesn't use the same set of external workspaces (like for example, a build for one of several targets in the project), external paths start breaking--and the red squiggles come out to indicate their displeasure. The temporary build sandbox, the execution root ("execroot") has gotten torn down and reconfigured, so some paths no longer exist. You can work around this by re-syncing, but it's annoying to have building break editing--feels like clearly undesirable behavior.

You can see that this problem is happening by opening, say, an NDK file in ASwB and looking at the header include paths. You can see the header include paths for a given file in Android Studio by "Show Compiler Info" (via Help→Find Action, aka CMD+SHIFT+A). You'll see paths containing `/private/var/tmp/_bazel_<username>/<hash>/execroot/__main__/external`, and verify that the symlinks those paths go through broken if you build some other part will get broken if a build that doesn't use the external workspace in question.

The fix: Those files live more durably inside `/private/var/tmp/_bazel_<username>/<hash>/external`, so the execroot-based paths should be rewritten to point there, just as we'd do for paths in the execroot that point into the normal workspace. That way you aren't relying on the symlinks in execroot that often get deleted out from under you!

Thanks for the work you do!
Chris
(ex-Googler)

P.S. This is parallel to a lingering Bug in Tulsi: https://github.com/bazelbuild/tulsi/issues/164 [Now resolved.]
P.P.S. If someone's taking a serious look at this, it'd be good to check that there isn't a parallel problem with generated files.

**Comment by cpsauer** [comment]
I've now fixed this in my working fork. The commit should be automatically backlinked above if the Bazel folk to use the code.

And if anyone using the plugin wants the fix (and others) the meantime, you can grab the plugin from:
 https://github.com/hedronvision/bazelbuild-intellij

**Comment by sgowroji** [comment]
Thank you @cpsauer   for sharing the above request. Could you Please share the PR's, I can arrange it for review.

**Comment by cpsauer** [comment]
Couple more notes before we decide that that PR fixes things:
1. The fix to #1256, linked above, looks *dicy*. Seriously, that blaze-out patch just covered bugs, imo, since you could easily have a deeply nested directory named external and it'll munch all those subdirectories. You want to use the regex I used here. We should merge this and have it call into this code, if it doesn't already.
2. I've got at least one other C++ project-breaking bug over in the fork I maintain. Please please take a peek and I'd be delighted to finally get to move on from that fork. https://github.com/hedronvision/bazelbuild-intellij

**Comment by cpsauer** [comment]
@sgowroji, is a review forthcoming on these? Checking in, since it's been a couple weeks.

**Comment by sgowroji** [comment]
Thank you @cpsauer for patience. We have shared the details to the engineer who will assist you soon on the above issue.

**Comment by cpsauer** [comment]
Friendly bump! Checking in since you'd asked for a PR for review and it's been a month.
Thanks for all you do.

[Btw, I also happened to be browsing the gazelle source, and [noted that they also handle external correctly, under output_base](https://github.com/bazelbuild/bazel-gazelle/blob/50455ffecf686e25e4b336328d8e35a3cf04a113/repo/repo.go#L50), as per the PR.

**Comment by Starkiller4011** [comment]
C++ user here looking for this to be implemented soonest so I can get rid of those annoying red squiggles. Thanks.

**Comment by cpsauer** [comment]
Maintainers: 
On labels, this is definitely a bug and not a feature request, IMO, since it's unintended behavior of existing functionality rather than new  functionality.
And please don't close after the resolution of #4063. Unresolved sub-issues in https://github.com/bazelbuild/intellij/issues/2766#issuecomment-1298104145

**Comment by cpsauer** [comment]
@Starkiller4011, heads that the general/C++ fix was just merged in.

[Others, still please don't close bc buggy code for Jars, Go, (and maybe others) still exists.]
